src/dynamic_fmt.c

```diff
@@ -404,6 +408,46 @@ char *RemoveHEX(char *output, char *input)
 	return output;
 }
 
+// Add $HEX$ encoding for salts where needed.  For canonicalized pot storage, we encode anything
+// containing unprintables or field separators, but never anything else
+static char *AddHEX(char *ciphertext) {
+	static char out[1024];
+	char *cp;
+
+	if (strncmp(ciphertext, "$dynamic_", 9) && strncmp(ciphertext, "@dynamic=", 9))
+		return ciphertext;  // not a dynamic format, so we can not 'fix' it.
+
+	// ok, see if this is salted:
+	if (!(cp = strchr(&ciphertext[1], ciphertext[0])) || !(cp = strchr(&cp[1], '$')))
+		return ciphertext;  // not a salted format.
+
+	if (!strncmp(cp, "$HEX$", 5))
+		return ciphertext;  // already HEX$
+
+	char *s = cp++;
+	int need_hex = 0;
+	if (cp[strlen(cp) - 1] == ' ' || cp[strlen(cp) - 1] == '\t' )
```

Comment:
I don't see whether/why `strlen(cp)` would be at least 1 here. Looks unsafe to me.